### PR TITLE
[jshint] add types for jshint

### PR DESCRIPTION
**Pull request by aminya**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/jshint/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
+// Type definitions for jshint 2.12
+// Project: http://jshint.com/
```
Comment (peterblazejewicz) [inline types/jshint/index.d.ts:2]:
```suggestion
// Project: https://github.com/jshint/jshint
```
this is optional, can we swap this to repo location? thx!

File: types/jshint/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+// Definitions by: Amin Yahyaabadi <https://github.com/me>
```
Comment (peterblazejewicz) [inline types/jshint/index.d.ts:3]:
```suggestion
// Definitions by: Amin Yahyaabadi <https://github.com/aminya>
```

File: types/jshint/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ *
+ * @param options The linting options to use when analyzing the source code.
+ * Type: an object whose property names are the desired options to use and whose property values are the configuration values for those properties.
+ * @example: `JSHINT(mySource, { undef: true });`
+ *
+ * @param predef variables defined outside of the current file; the behavior of this argument is identical to the globals linting option.
+ * Type: an object whose property names are the global variable identifiers and whose property values control whether each variable should be considered read-only
+ * @example: `JSHINT(mySource, myOptions, { jQuery: false });`
+ */
+export function JSHINT(source: string | string[], options?: LintOptions, predef?: Record<string, boolean>): LintError[];
```
Comment (peterblazejewicz) [inline types/jshint/index.d.ts:67]: Are you sure about the return type? I think it returns false/true and depending on the check state
one can iterate errors:
https://github.com/jshint/jshint/blob/master/src/cli.js#L495

File: types/jshint/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
+// Type definitions for jshint 2.12
+// Project: http://jshint.com/
+// Definitions by: Amin Yahyaabadi <https://github.com/me>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (peterblazejewicz) [inline types/jshint/index.d.ts:6]: why do you need these?

File: types/jshint/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+export interface LintError {
+    id: string;
+    raw: string;
+    code: string;
+    evidence: string;
+    line: number;
+    character: number;
+    scope: string;
+    reason: string;
+    [key: string]: any;
```
Comment (peterblazejewicz) [inline types/jshint/index.d.ts:17]: Isn't that limited to nullable `a`,`b`,`c`,`d` only?
https://github.com/jshint/jshint/blob/master/src/jshint.js#L407-L419

**Comment by peterblazejewicz** [comment]
@aminy pls review my comments, thx!

File: types/jshint/index.d.ts (hunk 4 above)
Comment (aminya) [inline types/jshint/index.d.ts:6]:
```suggestion
/* eslint-disable @typescript-eslint/no-unused-vars */
```

Comment (aminya) [reply types/jshint/index.d.ts:6]:
```suggestion
/* eslint-disable @typescript-eslint/no-unused-vars */
```

File: types/jshint/index.d.ts (hunk 5 above)
Comment (aminya) [reply types/jshint/index.d.ts:16]: I thought this a, b, c, d is arbitrary! But it seems you are right
```suggestion
  a?: any
  b?: any
  c?: any
  d?: any
```

File: types/jshint/index.d.ts (hunk 3 above)
Comment (aminya) [reply types/jshint/index.d.ts:69]: Yes, that's right.
```suggestion
export function JSHINT(source: string | string[], options?: LintOptions, predef?: Record<string, boolean>): boolean;
```

**Comment by aminya** [comment]
> @aminy pls review my comments, thx!

@peterblazejewicz Addressed all the comments!

**Comment by peterblazejewicz** [comment]
@aminya thx and congrats!

**Pull request merged**